File: MyLearning/DsaPractice/Graph/BellmanFordAlgorithm.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace DsaPractice.Graph
{
    /* 
     *  Bellman-Ford technically revisits all edges each time.
        But distances to nodes with the shortest path using k edges are effectively “finalized” after the k-th iteration.
        Further relaxations don’t change those distances unless there’s a negative cycle 
       (which Bellman-Ford would catch in the final pass).
     */
    internal class BellmanFordAlgorithm
    {
        public int[] BellmanFord(int V, int[,] edges, int src)
        {
            int nodes = V;

            int[] distances = [.. Enumerable.Range(0, nodes).Select(i => i == src ? 0 : int.MaxValue)];

            int rows = edges.GetLength(0);

            for (int i = 0; i < nodes; i++)
            {
                for (int j = 0; j < rows; j++)
                {
                    int from = edges[j, 0];
                    int to = edges[j, 1];
                    int weight = edges[j, 2];

                    int newDistance = distances[from] + weight;
                    int existingDistance = distances[to];

                    if (distances[from] != int.MaxValue && newDistance < existingDistance)
                    {
                        if (i == nodes - 1) // negative cycle detected
                            return [];

                        distances[to] = newDistance;

                    }
                }
            }

            return distances;
        }

        public void Test()
        {
            // Number of vertices in the graph
            int V = 5;

            int[,] edges = 
                {
                { 1, 3, 2 },
                { 4, 3, -1 },
                { 2, 4, 1 },
                { 1, 2, 1 },
                { 0, 1, 5 }
                };

            int src = 0;

            int[] ans = BellmanFord(V, edges, src);

            foreach (var d in ans)
            {
                Console.Write(d == int.MaxValue ? "INF, " : d + ", ");
            }
        }
    }
}

File: MyLearning/PythonPractice/Graph/BellmanFordAlgorithm.py
from typing import Iterable, List, Tuple


def bellman_ford(v: int, edges: Iterable[Tuple[int, int, int]], src: int) -> List[int]:
    dist = [float("inf")] * v
    dist[src] = 0
    edge_list = list(edges)
    for i in range(v):
        updated = False
        for frm, to, weight in edge_list:
            if dist[frm] == float("inf"):
                continue
            new_dist = dist[frm] + weight
            if new_dist < dist[to]:
                if i == v - 1:
                    return []
                dist[to] = new_dist
                updated = True
        if not updated:
            break
    return [int(d) if d != float("inf") else 2**31 - 1 for d in dist]
